src/upgrades/mongodb_upgrade.py

```diff
@@ -310,25 +399,69 @@ def are_replica_set_nodes_healthy(self, mongodb_config: MongoDBConfiguration) ->
 
     def is_cluster_able_to_read_write(self) -> bool:
         """Returns True if read and write is feasible for cluster."""
-        if self.charm.is_role(Config.Role.SHARD):
-            logger.debug("Cannot run read/write check on shard, must run via config-server.")
-            return False
-        elif self.charm.is_role(Config.Role.CONFIG_SERVER):
-            # TODO Future PR - implement node healthy check for sharded cluster
-            pass
-        else:
+        if self.charm.is_role(Config.Role.REPLICATION):
             return self.is_replica_set_able_read_write()
+        else:
+            return self.is_sharded_cluster_able_to_read_write()
+
+    def is_sharded_cluster_able_to_read_write(self) -> bool:
+        """Returns True if possible to write all cluster shards and read from all replicas."""
```

Comment:
In theory I like this idea, but I think that the MongoDB balancer will not break up the collection across shards until the collection gets much bigger. (i.e. until there is an uneven chunk distribution across shards, where a chunk is 64megabytes by default). We cannot force the balancer to distribute equally across all shards and replicas until the collection is quite large. 

I believe that what we have here is sufficient.

Reply:
up to @Mehdi-Bendriss

Reply:
I did some reading about this - I believe we can run this check without waiting for an automated rebalancing of the sharded collection (defined by the 64mb default).
The following steps should do it:
1. Get the count of the current shards, `shardsCount = 4`
2. Write data in an equally distributed manner (with balaned `shardKey` values, meaning: out of a 100 documents collection, we should get 25 unique documents on each of the 4 shards. Something along the lines of:
```
sh.shardCollection("testDb.test", { shardKey: 1 })
documents = [{"_id": i, "shardKey": i % shardsCount, "data": f"test-data_{i}"} for i in range(100)]
collection.insert_many(documents)
```
3. Verify the shard distribution - we should expect each shard to have 25 documents (given a 4 shards cluster):
```
distribution = collection.getShardDistribution()
# check the count for each shard
```
4. Verify that each replica within the replicaset of each shard contains the 25 documents and that the values correspond to the expected ones and are not sharded withe the other shards.

Reply:
I like this idea and I am grateful for your research here! It is certainly helpful 😄 , but we have limited time and I want to prioritise wrapping this up since I believe we will be finishing this 1w-ish later than anticipated (granted I've had a cumulative of 2W days off since the start of the 6mo cycle) I think that what we have here is sufficient for our needs , and the suggestion you both provided would only further improve the code ⭐ 

Can we save this as a "would be nice" for the end of the 24.10?

Reply:
So I ran those commands (they needed some small tweaks, which I will list here so that we can remember them in the future)
```
use admin
shards = db.runCommand({listShards: 1})["shards"]

# use the result from above to make this
let shards = [
    {"_id": "shard0"},
    {"_id": "shard1"}
];

let key_shards = [];
for (let i = 0; i < shards.length; i++) {
    key_shards.push([i, shards[i]._id]);
}

let repeated_key_shards = [];
key_shards.forEach(ks => {
    for (let i = 0; i < 25; i++) {
        repeated_key_shards.push(ks);
    }
});

let documents = repeated_key_shards.map((ks, i) => {
    return { _id: i, shardKey: ks[0], data: ks[1] };
});


shards = db.runCommand({listShards: 1})["shards"]
use testDb
 
let shards = [
    {"_id": "shard0"},
    {"_id": "shard1"}
];

let key_shards = [];
for (let i = 0; i < shards.length; i++) {
    key_shards.push([i, shards[i]._id]);
}

let repeated_key_shards = [];
key_shards.forEach(ks => {
    for (let i = 0; i < 25; i++) {
        repeated_key_shards.push(ks);
    }
});


sh.shardCollection("testDb.test", { shardKey: 1 })

let documents = repeated_key_shards.map((ks, i) => {
    return { _id: i, shardKey: ks[0], data: ks[1] };
});
db.test.insertMany(documents)

db.adminCommand( { flushRouterConfig: "test.myShardedCollection" } );
distribution = db.test.getShardDistribution()
```

and the result was:
```
{
  data: '2KiB',
  docs: 50,
  chunks: 1,
  'estimated data per chunk': '2KiB',
  'estimated docs per chunk': 50
}
---
Totals
{
  data: '2KiB',
  docs: 50,
  chunks: 1,
  'Shard shard1': [
    '100 % data',
    '100 % docs in cluster',
    '45B avg obj size on shard'
  ]
}
```

and according to the docs:
> <estDataPercent-x> is a calculated value that reflects, for each shard, the data size as the percentage of the collection's total data size, for example:

which means that even though we had the shard key specified all the data is still on Shard1 (which I assume to be the case due to the chunk size)